File: solutions/0/0/9/0/p0090_s0.c
/* Solution to Euler Problem 90: Cube Digit Pairs. */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

static int can_display(int *cube, int digit) {
    for (int i = 0; i < 6; i++) {
        if (cube[i] == digit) return 1;
    }
    if (digit == 6 || digit == 9) {
        for (int i = 0; i < 6; i++) {
            if (cube[i] == 6 || cube[i] == 9) return 1;
        }
    }
    return 0;
}

static int can_pair_display_all(int *cube1, int *cube2) {
    /* squares: 01,04,09,16,25,36,49,64,81 */
    int squares[9][2] = {
        {0,1},{0,4},{0,9},{1,6},{2,5},{3,6},{4,9},{6,4},{8,1}
    };
    for (int s = 0; s < 9; s++) {
        int a = squares[s][0];
        int b = squares[s][1];
        if (!((can_display(cube1, a) && can_display(cube2, b)) ||
              (can_display(cube1, b) && can_display(cube2, a)))) {
            return 0;
        }
    }
    return 1;
}

long long solve(int argc, char *argv[]) {
    /* Generate all C(10,6) = 210 combinations */
    int all_cubes[210][6];
    int cube_count = 0;

    for (int a = 0; a <= 4; a++)
    for (int b = a+1; b <= 5; b++)
    for (int c = b+1; c <= 6; c++)
    for (int d = c+1; d <= 7; d++)
    for (int e = d+1; e <= 8; e++)
    for (int f = e+1; f <= 9; f++) {
        all_cubes[cube_count][0] = a;
        all_cubes[cube_count][1] = b;
        all_cubes[cube_count][2] = c;
        all_cubes[cube_count][3] = d;
        all_cubes[cube_count][4] = e;
        all_cubes[cube_count][5] = f;
        cube_count++;
    }

    long long valid_arrangements = 0;
    for (int i = 0; i < cube_count; i++) {
        for (int j = i; j < cube_count; j++) {
            if (can_pair_display_all(all_cubes[i], all_cubes[j])) {
                valid_arrangements++;
            }
        }
    }
    return valid_arrangements;
}

/* Usage: ./file <kwarg>... [--runs=1] [--show]
 * Output: "<runs> <avg_seconds> <result>" */
int main(int argc, char *argv[]) {
    int runs = 1;

    char **solve_argv = malloc((size_t)argc * sizeof(char *));
    if (!solve_argv) {
        fprintf(stderr, "runner: out of memory\n");
        return 1;
    }
    int solve_argc = 0;
    solve_argv[solve_argc++] = argv[0];

    for (int i = 1; i < argc; i++) {
        if (argv[i][0] == '\0') continue;
        if (strncmp(argv[i], "--runs=", 7) == 0) {
            int r = atoi(argv[i] + 7);
            if (r >= 1) runs = r;
            continue;
        }
        if (strcmp(argv[i], "--show") == 0) continue;
        solve_argv[solve_argc++] = argv[i];
    }

    long long result = 0;
    double total = 0.0;
    int rc = 0;
    int has_result = 0;

    for (int r = 0; r < runs; r++) {
        struct timespec t0, t1;
        clock_gettime(CLOCK_MONOTONIC, &t0);
        long long cur = solve(solve_argc, solve_argv);
        clock_gettime(CLOCK_MONOTONIC, &t1);
        total += (double)(t1.tv_sec - t0.tv_sec)
               + (double)(t1.tv_nsec - t0.tv_nsec) * 1e-9;
        if (has_result && cur != result) {
            fprintf(stderr, "Expected consistent result, got %lld previous result=%lld\n",
                    cur, result);
            rc = 1;
        }
        result = cur;
        has_result = 1;
    }

    free(solve_argv);
    printf("%d %.17g %lld\n", runs, total / (double)runs, result);
    return rc;
}

File: solutions/0/0/9/0/p0090_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 90: Cube Digit Pairs [Level 9]. """
from __future__ import annotations

import itertools

from solver.runners import runner


@runner.main
def solve(*args: str) -> str:
    """Brute-force over all unordered pairs of the C(10,6)=210 six-digit cubes; O(210^2 * 9)."""
    squares: list[tuple[int, int]] = [(int((i_sq := f"{i * i:02d}")[0]), int(i_sq[1])) for i in range(1, 10)]

    def can_display(cube_digits: tuple[int, ...], digit: int) -> bool:
        """True if the cube shows the digit, treating 6 and 9 as interchangeable via the flip rule."""
        return digit in cube_digits or (digit in [6, 9] and (6 in cube_digits or 9 in cube_digits))

    def can_pair_display_all(cube1: tuple[int, ...], cube2: tuple[int, ...]) -> bool:
        """True if every square's two digits can be shown, with either cube supplying the tens digit."""
        for first_digit, second_digit in squares:
            if not (
                can_display(cube1, first_digit)
                and can_display(cube2, second_digit)
                or (can_display(cube1, second_digit) and can_display(cube2, first_digit))
            ):
                return False
        return True

    all_cubes: list[tuple[int, ...]] = list(itertools.combinations(range(10), 6))
    cube_count: int = len(all_cubes)
    valid_arrangements: int = 0
    # Triangular iteration (j from i) counts each unordered pair once and admits self-pairs.
    for i in range(cube_count):
        for j in range(i, cube_count):
            if can_pair_display_all(all_cubes[i], all_cubes[j]):
                valid_arrangements += 1
    return str(valid_arrangements)


if __name__ == "__main__":
    raise SystemExit(solve())
